### PR TITLE
Update SQL DecoderContext to build a global module and top block. 

### PR DESCRIPTION
**Pull request by markww**
Before, we were adding `Operation`s to an ad-hoc structure involving
a `std::vector` of `Operation`s. This change causes us to instead build
up a global `Module` and a top-level `Block` to hold these (and other)
IR nodes.

### CODE REVIEW

**Comment by markww** [comment]
Note: the diff will be a little messed up until `forbid_double_building_block_builder` is merged in.

**Comment by bgogul** [comment]
@markww can you merge all the dependent PRs and rebase this one? It is hard to see the changes relevant for this PR alone.

**Comment by markww** [comment]
Sure thing.

**Comment by markww** [comment]
Alright, I've cleaned this up, should be much easier to read now.

File: src/frontends/sql/decoder_context_test.cc [hunk 1]
```diff
@@ -31,14 +31,23 @@ using testing::UnorderedElementsAre;
...
-using ir::value::StoredValue;
 using ir::value::OperationResult;
+using ir::value::StoredValue;
+
+TEST(DecoderContextTest, BuildTopLevelBlock) {
+  IRContext context;
+  DecoderContext decoder_context(context);
+
+  const Block &block = decoder_context.BuildTopLevelBlock();
+  EXPECT_EQ(block.module(), &decoder_context.global_module());
```
Comment (bgogul) [inline src/frontends/sql/decoder_context_test.cc:49]: Didn't we change `block.module()` to `block.parent_module()`?

File: src/frontends/sql/decoder_context.h [hunk 2]
```diff
@@ -36,30 +36,32 @@ class DecoderContext {
...
 
   // Register a value as associated with a particular id. This stores that
   // value and provides a reference to its canonical value.
   const ir::Value &RegisterValue(uint64_t id, ir::Value value) {
     auto insert_result =
         id_to_value.insert({id, std::make_unique<ir::Value>(std::move(value))});
     CHECK(insert_result.second)
-      << "id_to_value map has more than one value associated with the id "
-      << insert_result.first->first << ".";
+        << "id_to_value map has more than one value associated with the id "
```
Comment (bgogul) [inline src/frontends/sql/decoder_context.h:51]: nit: why do we have these formatting changes now? Have  you started using `clang-format` recently?

**Review by bgogul — APPROVED** [review]

File: src/frontends/sql/decoder_context.h (hunk 2 above)
Comment (markww) [reply src/frontends/sql/decoder_context.h:51]: I have. I didn't have it set up properly before, and now I'm frequently using it to keep my formatting standard.

File: src/frontends/sql/decoder_context_test.cc (hunk 1 above)
Comment (markww) [reply src/frontends/sql/decoder_context_test.cc:49]: Yes. I changed the field but forgot the getter :(

Comment (markww) [reply src/frontends/sql/decoder_context_test.cc:49]: I'll do a follow-on PR to fix that.

**Pull request merged**